### PR TITLE
CASSANDRA-18521: Unify CQLTester#waitForIndex and SAITester#waitForIndexQueryable

### PR DESCRIPTION
**Pull request by adelapena**

### CODE REVIEW

File: test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexOnMapEntriesTest.java [hunk 1]
```diff
@@ -230,7 +230,7 @@ public void testShouldRejectQueriesForNullEntries() throws Throwable
     public void testShouldTreatQueriesAgainstFrozenMapIndexesAsInvalid() throws Throwable
     {
         createTable("CREATE TABLE %s (k TEXT PRIMARY KEY, v FROZEN<MAP<TEXT, TEXT>>)");
-        createIndex("CREATE INDEX ON %s(FULL(V))");
+        createIndexAsync("CREATE INDEX ON %s(FULL(V))");
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexOnMapEntriesTest.java]: nit: It doesn't affect correctness here, but would it be better to use `createIndex()` just to make sure that there's no lingering index build task active when the test completes? (i.e. If the default 2i changes later...)

Comment (maedhroz) [reply test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexOnMapEntriesTest.java]: This happens in a number of other tests, so I suppose I won't comment on every single one, but we could do a quick audit if we want to change this in general...

File: test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java [hunk 2]
```diff
@@ -1060,9 +1057,8 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
     public void testReadOnlyIndex() throws Throwable
     {
         // On successful initialization both reads and writes go through
-        String tableName = createTable("CREATE TABLE %s (pk int, ck int, value int, PRIMARY KEY (pk, ck))");
+         createTable("CREATE TABLE %s (pk int, ck int, value int, PRIMARY KEY (pk, ck))");
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java]:
```suggestion
        createTable("CREATE TABLE %s (pk int, ck int, value int, PRIMARY KEY (pk, ck))");
```

File: test/unit/org/apache/cassandra/cql3/CQLTesterTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        testGetCreateIndexName("k", "idx", "CREATE INDEX idx ON k.t (c)");
+        testGetCreateIndexName(KEYSPACE, "idx", "CREATE INDEX idx ON t (c)");
+        testGetCreateIndexName(KEYSPACE, "idx", "CREATE INDEX idx ON %s (c)");
+
+        // custom index with explicit name
+        testGetCreateIndexName("k", "idx", "CREATE CUSTOM INDEX idx ON k.t (c) USING 'class'");
+        testGetCreateIndexName(KEYSPACE, "idx", "CREATE CUSTOM INDEX idx ON t (c) USING 'class'");
+        testGetCreateIndexName(KEYSPACE, "idx", "CREATE CUSTOM INDEX idx ON %s (c) USING 'class'");
+
+        // with a unqouted case-sensitive index name
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTesterTest.java]:
```suggestion
        // with an unqouted case-sensitive index name
```

File: test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java [hunk 4]
```diff
@@ -176,7 +176,7 @@ public void createIndexes(Executor tester) throws Throwable
             if (!skipColumns.contains(column.left))
             {
                 executeLocalIndexed(tester, String.format(template, column.left, indexedTable, column.left));
-                tester.waitForIndexQueryable(KEYSPACE, indexedTable);
+                tester.waitForTableIndexesQueryable(KEYSPACE, indexedTable);
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java]: I think the original intention of this and the call on 304 was to wait for the single column index just created. In that case, we need to either use `waitForIndexQueryable(keyspace, indexName)` or move the `waitForTableIndexesQueryable()` outside the loop. The former might be better isolated. WDYT?

Comment (maedhroz) [reply test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java]: Might require a change to `Executor`, but a small one...

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 5]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
+
+    protected boolean areTableIndexesQueryable()
     {
-        long start = currentTimeMillis();
-        SecondaryIndexManager indexManager = getCurrentColumnFamilyStore(keyspace).indexManager;
+        return areTableIndexesQueryable(KEYSPACE, currentTable());
+    }
 
-        while (true)
+    protected boolean areTableIndexesQueryable(String keyspace, String table)
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]: The test failures we're seeing lead me to believe that we should actually have this method return the set of indexes that haven't yet become queryable. That might give us some insight into what's going on if the assertion times out. (In other words, we can use `assertEquals(<empty set>, areTableIndexesQueryable(...))` in `waitForTableIndexesQueryable()` and get more information on failure to help debug.)

Comment (maedhroz) [reply test/unit/org/apache/cassandra/cql3/CQLTester.java]: https://app.circleci.com/pipelines/github/adelapena/cassandra/2892/workflows/837b186e-6cd9-442c-baaf-8bc067f56b21/jobs/43816/tests#failed-test-0

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 6]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
     {
         String formattedQuery = formatQuery(keyspace, query);
-        return createFormattedIndex(formattedQuery);
+        Pair<String, String> qualifiedIndexName = createFormattedIndex(keyspace, formattedQuery);
+        waitForIndexQueryable(qualifiedIndexName.left, qualifiedIndexName.right);
+        return qualifiedIndexName.right;
+    }
+
+    /**
+     * Creates a secondary index, without waiting for it to become queryable.
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]:
```suggestion
     * Creates a secondary index without waiting for it to become queryable.
```

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 7]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
+     * @param query the index creation query
+     * @return the name of the created index
+     */
+    protected String createIndexAsync(String query)
+    {
+        return createIndexAsync(KEYSPACE, query);
+    }
+
+    /**
+     * Creates a secondary index, without waiting for it to become queryable.
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]:
```suggestion
     * Creates a secondary index without waiting for it to become queryable.
```

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 8]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
+    public void waitForTableIndexesQueryable(String table)
+    {
+        waitForTableIndexesQueryable(KEYSPACE, table);
     }
 
     /**
-     * Index creation is asynchronous, this method searches in the system table IndexInfo
-     * for the specified index and returns true if it finds it, which indicates the
-     * index was built. If we haven't found it after 5 seconds we give-up.
+     * Index creation is asynchronous, this method waits until all the indexes in the specified table are queryable.
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]:
```suggestion
     * Index creation is asynchronous. This method waits until all the indexes in the specified table are queryable.
```

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 9]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
-                break;
+    public void waitForIndexQueryable(String index)
+    {
+        waitForIndexQueryable(KEYSPACE, index);
+    }
 
-            Thread.sleep(10);
-        }
+    /**
+     * Index creation is asynchronous, this method waits until the specified index is queryable.
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]:
```suggestion
     * Index creation is asynchronous. This method waits until the specified index is queryable.
```

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 10]
```diff
@@ -1031,122 +1034,222 @@ protected void alterTableMayThrow(String query) throws Throwable
...
+    }
 
-        return indexCreated;
+    protected void waitForIndexBuilds(String index)
+    {
+        waitForIndexBuilds(KEYSPACE, index);
     }
 
     /**
      * Index creation is asynchronous, this method waits until the specified index hasn't any building task running.
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]:
```suggestion
     * Index creation is asynchronous. This method waits until the specified index hasn't any building task running.
```

File: test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexOnMapEntriesTest.java (hunk 1 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexOnMapEntriesTest.java]: Changed to `createIndex`. I have found many of this, most uses of `createAsync` are for indexes that use either an injection or a custom implementation that will fail or block the build.

File: test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java (hunk 2 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java]: Done.

File: test/unit/org/apache/cassandra/cql3/CQLTesterTest.java (hunk 3 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/CQLTesterTest.java]: Done.

File: test/unit/org/apache/cassandra/cql3/CQLTester.java (hunk 5 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/CQLTester.java]: I have replaced it by a `getNotQueryableIndexes` method.

File: test/unit/org/apache/cassandra/cql3/CQLTester.java (hunk 6 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/CQLTester.java]: Done.

File: test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java (hunk 4 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java]: Yep, I've changed it to use `waitForIndexQueryable` and removed some minor code duplication.

File: test/unit/org/apache/cassandra/cql3/CQLTester.java (hunk 5 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/cql3/CQLTester.java]: It seems the failure can be reproduced on the base feature branch too: https://app.circleci.com/pipelines/github/adelapena/cassandra/2894/workflows/71d2d463-2780-496d-a6fd-8c530768ae1b/jobs/44121/tests

Comment (maedhroz) [reply test/unit/org/apache/cassandra/cql3/CQLTester.java]: Yeah, I'm fine merging w/ `getNotQueryableIndexes()` in place and then we'll at least get a more helpful error when this fails again.

File: test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java [hunk 11]
```diff
@@ -169,14 +169,22 @@ public void createTables(Executor tester)
 
     public void createIndexes(Executor tester) throws Throwable
     {
-        String template = "CREATE CUSTOM INDEX ndi_%s_index_%s ON %%s (%s) USING 'StorageAttachedIndex'";
+        createIndexes(tester, columns);
+    }
+
+    protected void createIndexes(Executor tester, List<Pair<String, String>> columns) throws Throwable
+    {
+        String indexNameTemplate = "ndi_%s_index_%s";
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java]: nit: Ha, missed this earlier. "ndi" => "sai" I guess

**Review by maedhroz — APPROVED** [review]

File: test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java (hunk 11 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/index/sai/cql/BaseDataModel.java]: Fixed, this one and a similar one in `OperationTest.STATIC_CF_NAME`.

**Pull request closed without merging**